### PR TITLE
Added snapcraft configuration file to build a snap

### PR DESCRIPTION
**Pull request by r-f-g**
- snapcraft.yaml provide configuration to build the prometheus-bind-exporter
 - this snap should land first in the edge channel [1]

 ---
 [1]: https://snapcraft.io/prometheus-bind-exporter

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+name: prometheus-bind-exporter
+version: git
+title: Prometheus bind exporter
+summary: Export BIND (named/dns) v9+ service metrics to Prometheus.
+description: Export BIND (named/dns) v9+ service metrics to Prometheus.
+
+grade: stable
+confinement: classic
```
Comment (auria) [inline snap/snapcraft.yaml:8]: This needs to be `strict`. Classic "confinements" (unconfined, really) is supported for applications who can't work on strictly confined snap mode (snap interfaces missing or other reasons).

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+name: prometheus-bind-exporter
+version: git
```
Comment (auria) [inline snap/snapcraft.yaml:2]: When I run `snapcraft`, which runs all the stages (pull, build, stage, prime), it fails with the following error:

```
Building prometheus-bind-exporter 
+ snapcraftctl build
+ go mod download
+ go install -p 8 -ldflags -linkmode=external ./...
Staging prometheus-bind-exporter 
+ snapcraftctl stage
Priming prometheus-bind-exporter 
+ snapcraftctl prime
Determining the version from the project repo (version: git).
fatal: not a git repository (or any of the parent directories): .git
Run the same command again with --debug to shell into the environment if you wish to introspect this failure.
```

When I changed the version to `0.4.git`, it worked. I think snaps should always be built against tagged versions, though. See [this example](https://github.com/canonical/juju-verify/blob/a06f21847c6d76eecb3dee171f80d4a0b828ce4f/snap/snapcraft.yaml#L36) on how to set a version via a build override.

File: README.md [hunk 3]
```diff
@@ -26,6 +26,12 @@ docker pull prometheuscommunity/bind-exporter:v0.3.0
 docker run -d --network host prometheuscommunity/bind-exporter:v0.3.0
 ```
 
+### Install with Snap
+
+```bash
+snap install prometheus-bind-exporter --classic
```
Comment (auria) [inline README.md:32]: This should be installed without `--classic`.

**Review by auria — CHANGES_REQUESTED** [review]
Please find some comments inline.

The snap needs to be strictly confined, and running as a service, on a configured port. This is needed so Prometheus can connect to the exporter via HTTP at any time.

The [Prometheus snap](https://git.launchpad.net/prometheus-snap/tree/snapcraft.yaml) uses [a shell script as a wrapper](https://git.launchpad.net/prometheus-snap/tree/snap_config_wrapper). The snap "app" is of type "daemon" so whenever the snap is executed, a service will be started. The systemd service will be called `snap.<snapname>`.

Please look for snaps for other exporters, like the prometheus-openstack-exporter of the prometheus-ceph-exporter.

**Comment by SuperQ** [comment]
We don't support snap in Prometheus.

**Comment by auria** [comment]
@SuperQ we are planning to support a snap version of this project. Would it be preferable to track the effort in a repo maintained by us? Thanks.

**Comment by SuperQ** [comment]
Yes, if you plan to support it, you should do this separately.

**Pull request closed without merging**